clang/include/clang/Basic/Attr.td

```diff
@@ -1226,6 +1226,14 @@ def SYCLKernel : InheritableAttr {
   let Documentation = [SYCLKernelDocs];
 }
 
+def SYCLSpecialClass: InheritableAttr {
+  let Spellings = [Clang<"sycl_special_class", /*allowInC =*/0>];
+  let Subjects = SubjectList<[CXXRecord]>;
+  let LangOpts = [SYCLIsDevice, SilentlyIgnoreSYCLIsHost];
+  let Documentation = [Undocumented];
+  let Args = [IdentifierArgument<"ClassType">];
```

Comment:
@Naghasan Thanks for the review. This is an attribute to be applied to accessor, stream and sampler classes only.  
After a review comment, the doc  was removed since this is an attribute  not meant to be "seen" by users (although it resides in a header, therefore will be seen and expecting that users will eventually use it for classes other than the 3 mentioned). The goal is to upstream this to community, I will put back the doc.

Reply:
We might want to consider making this an `EnumArgument` instead so that we list the specific identifiers that are allowed. Then users will get a diagnostic if they try to use `[[clang::sycl_special_class(steram)]]` accidentally.

Reply:
If we  don't need to identify special classes individually, we may be able to do away with the argument. See: https://github.com/intel/llvm/pull/3892#discussion_r654630296.

Reply:
>If we don't need to identify special classes individually

This depends on how the various isXYZSpecialClass functions are used right now. If the only reason to identify these classes is the special handling when we generate openCL kernel, we could probably just do this based on __init/__finalize. Off the top of my head, I do not know if these functions are used for other purposes though.

Reply:
@AaronBallman That's how it was written initially using an enum. I think we need to decide how this attribute is going to be used. Is there a chance that it can be extended to other class kind? The goal of this attribute is to mark the classes that need special handling with this attribut instead of use their kind to recognize them. Since these are the only 3 classes that have the __init/finalize methods then no need to enum the kind. @Naghasan your thoughts? I think you were the one who proposed the recognition of these classes by the init/finalize method? Isn't this also related to what is going to be the use of this attribute? A general one or internal?

Reply:
I personally prefer the enum option to enumerate the kind class that are allowed and that would automatically restrict the user to use this attribute with other classes and would potentially allow more kind to be added if needed.

Reply:
Left is as is for now. Will put back the enum when/if there is consensus.

Reply:
FWIW as a drive-by, I also prefer the enum.  There are slight differences between how an accessor/sampler/etc is used if I recall correctly, and we'll want to continue 'knowing' this in the future.  We have a bunch of uses of `isSYCLAccessorType` that all should be removed thanks to this, and we can't know this unless it has one of a fixed set of values here.

Reply:
Actually... id further posit that the parameter itself is worthless unless we make it an enum.

Reply:
The parameter is here to differentiate between the accessor and sample classes since they both have __init method.

Reply:
Do we really need to KNOW the difference between `accessor` and `sample` for that?   Could they not both be a generic 'special' type?

If the answer to those is "We need to know the difference", these should be enum, since it is a fixed list.  If the answer to those is "no, we don't need to know the difference", then there should be no parameter at all.  An identifier parameter does not seem otherwise particularly useful.

Reply:
@elizabethandrews  -- were you suggesting that perhaps we could pull the identifier for the attribute from the name of the class the attribute is written on? e.g., `struct __attribute__((sycl_special_class)) stream {};`

Reply:
> 
> 
> @elizabethandrews -- were you suggesting that perhaps we could pull the identifier for the attribute from the name of the class the attribute is written on? e.g., `struct __attribute__((sycl_special_class)) stream {};`

I think that makes this attribute worthless, right?  Since that is what we're doing, other than having an attribute.  Currently we are directly recognizing the name of the types directly, so this would just be to limit that same name-recognition to the cases where the attribute is present.  I think that is not a better situation.

The enum values should be the list of ones we care to differentiate based on behavior, plus a 'generic' that has 'good enough behavior for most cases'.

Reply:
IIRC our previous discussions on this attribute, @Naghasan said that we can design special types so that there should be no difference between them in the context of SYCL kernel translation to SPIR-V kernel. If so, the compiler part could be simpler and handle all such types uniformly, but there might be additional common requirements for the special types definitions (e.g. must have `__init` and `__finalize` member functions).

Reply:
> Do we really need to KNOW the difference between `accessor` and `sample` for that? Could they not both be a generic 'special' type?
> 
> If the answer to those is "We need to know the difference", these should be enum, since it is a fixed list. If the answer to those is "no, we don't need to know the difference", then there should be no parameter at all. An identifier parameter does not seem otherwise particularly useful.

Not sure. Depending on the kind of the class in KernelObjVisitor::visitField we call a different function for accessor and sampler. If we remove the parameter the sampler class will hit the first test and call handleSyclAccessorType. If they are supposed to do the same thing, then that would work. Otherwise I think we need an enum.

Reply:
> 
> 
> > Do we really need to KNOW the difference between `accessor` and `sample` for that? Could they not both be a generic 'special' type?
> > If the answer to those is "We need to know the difference", these should be enum, since it is a fixed list. If the answer to those is "no, we don't need to know the difference", then there should be no parameter at all. An identifier parameter does not seem otherwise particularly useful.
> 
> Not sure. Depending on the kind of the class in KernelObjVisitor::visitField we call a different function for accessor and sampler. If we remove the parameter the sampler class will hit the first test and call handleSyclAccessorType. If they are supposed to do the same thing, then that would work. Otherwise I think we need an enum.

Yep, thats basically the question here.  That said, I am positive that they both have different behaviors in the integration header (accessor is its own type!), so I think we actually DO need a parameter.

Reply:
> > @elizabethandrews -- were you suggesting that perhaps we could pull the identifier for the attribute from the name of the class the attribute is written on? e.g., `struct __attribute__((sycl_special_class)) stream {};`
> 

No I was asking if we need to identify special SYCL classes individually. As in, if it is required to identify `stream ` vs `accessor`. If the only use of this attribute is to denote which classes need special handling, we may not need to differentiate between individual classes as long as all special types are handled the same way.

>>If the answer to those is "We need to know the difference", these should be enum, since it is a fixed list. If the answer to those is "no, we don't need to know the difference", then there should be no parameter at all. An identifier parameter does not seem otherwise particularly useful.

@erichkeane perhaps describes it better here.

Reply:
Ah, maybe I am confused? I was assuming there was a 1:1 correlation between these special classes and the types in the SYCL headers. e.g., there's one `stream` class and one `accessor` class, and other things inherit from those base classes if they want to. So my thinking was the attribute would live on the base class (say `stream`) to identify it as a special stream class thing, and thus we wouldn't need to put the identifier into the attribute at all. e.g.,
```
namespace sycl {
namespace __detail {
namespace whatever {
struct __attribute__((sycl_special_class)) stream {};
}
}
}
```
would be exactly the same as:
```
namespace sycl {
namespace __detail {
namespace whatever {
struct __attribute__((sycl_special_class(stream))) stream {};
}
}
}
```
Are you saying that the usage is more like?
```
namespace sycl {
namespace __detail {
namespace whatever {
struct __attribute__((sycl_special_class(stream))) brook {};
struct __attribute__((sycl_special_class(stream))) river {};
struct __attribute__((sycl_special_class(stream))) creek {};
}
}
}
```

Reply:
> 
> 
> Ah, maybe I am confused? I was assuming there was a 1:1 correlation between these special classes and the types in the SYCL headers. e.g., there's one `stream` class and one `accessor` class, and other things inherit from those base classes if they want to. So my thinking was the attribute would live on the base class (say `stream`) to identify it as a special stream class thing, and thus we wouldn't need to put the identifier into the attribute at all. e.g.,
> 
> ```
> namespace sycl {
> namespace __detail {
> namespace whatever {
> struct __attribute__((sycl_special_class)) stream {};
> }
> }
> }
> ```
> 
> would be exactly the same as:
> 
> ```
> namespace sycl {
> namespace __detail {
> namespace whatever {
> struct __attribute__((sycl_special_class(stream))) stream {};
> }
> }
> }
> ```
> 
> Are you saying that the usage is more like?
> 
> ```
> namespace sycl {
> namespace __detail {
> namespace whatever {
> struct __attribute__((sycl_special_class(stream))) brook {};
> struct __attribute__((sycl_special_class(stream))) river {};
> struct __attribute__((sycl_special_class(stream))) creek {};
> }
> }
> }
> ```

No, your use example is correct.  However, we ALREADY identify stream/accessor by name, and it is somewhat fragile.  The intent of the attribute is to stop having to investigate the name of the class to figure out what it is.

Additionally, adding new 'classes' of these is difficult, even though they do pretty much exactly what the others do. 

So the idea here is to permit the library to add a new class-type that WORKS like a `stream` or `accessor` (or just has an init and finalize, aka, generic!) without involving the CFE again.

I believe the 'open' is two decide whether it is actually important that we know it is '`stream` or `accessor`, or if we could just treat them all identically/generically.

Reply:
> No, your use example is correct. However, we ALREADY identify stream/accessor by name, and it is somewhat fragile. The intent of the attribute is to stop having to investigate the name of the class to figure out what it is.

I've always understood that the fragility is because we need to care about far too much of the name -- like the namespaces its within, whether it's a template, etc. I don't understand what's fragile about "the class is named `stream` so this is a stream type" -- do you think we're going to rename `stream` to something else (and break all the users of the API), but still need to know it's a stream object without changing the FE?

> So the idea here is to permit the library to add a new class-type that WORKS like a stream or accessor (or just has an init and finalize, aka, generic!) without involving the CFE again.

As in:
```
struct river : stream {}; // This is a new kind of stream?
```
I don't see why the FE would have to change for that. Put a different way -- if it's a new "kind" of stream or accessor, I'm imagining there's a type relationship between the new type and an existing stream/accessor type, so there'd be inheritance involved...

Reply:
I had a phone call with @erichkeane so we could get on the same page. Here's my summary of what we covered and what I now understand:

* It's not clear an attribute argument is needed because it's not clear that the FE *cares* about a distinction between the "kinds" of special classes. The FE currently cares, but we're not certain that's actually required. This is an open research question.
* If the FE does not need to distinguish between kinds of special classes, the attribute argument isn't necessary at all.
* If the FE does need to distinguish between kinds of special classes, there are two approaches with different tradeoffs:
* * We could figure out the kind based on the name of the class the attribute is applied to (no attr arg needed)
* * We could make the user explicitly specify the kind as an attribute argument
* * Either one of these two approaches would work, but there's a stronger preference for an explicit attribute argument because that enables more potential use cases (such as a "generic" special class kind on a class named `stream` in some detail namespace for something that is not actually a stream special class kind).

So I think I'm now on the same page and agree with Erich that either the attribute argument is unnecessary because the FE doesn't need to make a distinction between special classes or it needs an enum argument to explicitly specify the kind.

Reply:
@AaronBallman I'm wondering how we can be convince ourselves (meaning the FE) that there is no need to handle these classes differently? Is there a way to know/find that now? Meanwhile should this patch get back to the enum? Or wait until we know more research is done?

Reply:
> @AaronBallman I'm wondering how we can be convince ourselves (meaning the FE) that there is no need to handle these classes differently? Is there a way to know/find that now? 

As I mentioned in this comment https://github.com/intel/llvm/pull/3892#discussion_r657244156,  I think you will need to look at how the various "isXYZSpecialClass" functions are used now.

Reply:
@zahiraam I'd probably start with a code inspection of SemaSYCL.cpp to see how we use the distinction currently. My conversation with Erich led me to believe that the current "is this a special class of type <X>" logic there grew organically and that it wasn't clear that the "of type <X>" part was actually necessary in all cases. @erichkeane do you have other suggestions or thoughts?

Reply:
I'd prefer that we have done the research in advance of this feature.

Reply:
It is my understanding that when this attribute was originally proposed (I think last year), the special classes were all handled in a similar manner, or they were going to be redesigned that way. The purpose of this attribute (at that time) was to let the compiler know this is a SYCL special type and thus required the special processing with `__init` and `__finalize`. The assumption was that all special types would be handled in a similar manner with` __init` and `__finalize`.

I think the first step here now is to check the `handleXYZSpecialType()` functions in `SemaSYCL.cpp ` and see if they all do the same thing. (By XYZ I mean accessor, sampler, etc). 

I would also suggest looking at all calls of `isXYZSpecialClass()` to confirm these are only used to call handlers. If these are used in other areas of `SemaSYCL.cpp` we need to know why that unique identification of `XYZ ` was required.

Reply:
More research obviously has to be done, but I THINK our int-header generates some different 'types' based on accessor/sampler/stream.  If this is the case, we need to talk to the RT time to figure out if that is necessary, or if 'special type' is enough.  I suspect THAT is going to be the easiest way to do said research.  If they NEED to know, then we do too.

Reply:
At the moment, I think there is some difference when it comes to inheriting from these classes as well. IIRC this is however a bug that needs to be fixed.

Reply:
@erichkeane Who can be pulled from the RT in this discussion? I think looking at the SemaSYCL.cpp and the calls to the special functions type handling is not going to be enough.

Reply:
> , but I THINK our int-header generates some different 'types' based on accessor/sampler/stream. 

You're right. I just took a quick look We have  kind_accessor, kind_sampler, and kind_stream. 

I also just remembered the opt-report emits different information for each of these types but I think that can be changed to be generic I think

Reply:
@zahiraam : @romanovvlad  is usually who I talk to.

Reply:
I made a quick change to funnel all 3 classes into handleSyclAccessorType function in kernelObjVisitor::visitField. 
The stream class crashes in SyclKernelDeclCreator::handleSpecialType at the call of handleAccessoryPropertyList because it expects a templated class, and stream class is not templated.
The sampler class crashes SyclKernelHaderCreator::handleSyclAccessor  for the same reason. It is expecting a templated class as argument and sampler class is not templated.
Is that convincing enough to keep the attribute  and its argument? I am wondering now why is accessor templated and not the other 2!

Reply:
i'm not familiar with the `accessor-property-list`... First, should that be its own 'type' here?  That is, should it have its own entry in the enum? What special handling do we do for that type?

Reply:
The property_list is only used for the accessor class. Sampler and stream classes don’t have a property_list. It is used to gather information about the kernel arguments. When appropriate the attributes  SYCLIntelBufferLocation and/or RestrictAttr are added to the argument. 
So for example:

template <int I>
using buffer_location = cl::sycl::INTEL::property::buffer_location::instance<I>;
class AccessorBase {
  int A;
public:
  cl::sycl::accessor<int, 1, cl::sycl::access::mode::read_write,
                     cl::sycl::access::target::global_buffer,
                     cl::sycl::access::placeholder::false_t,
                     cl::sycl::ONEAPI::accessor_property_list<buffer_location<2>>>
  acc;
};

Or

int main() {
  cl::sycl::accessor<int, 1, cl::sycl::access::mode::read_write,
                     cl::sycl::access::target::global_buffer,
                     cl::sycl::access::placeholder::false_t,
                     cl::sycl::ONEAPI::accessor_property_list<
                         cl::sycl::ONEAPI::property::no_alias::instance<true>>>
      accessorA;
…
}

In the first case _arg_acc will get the SYCLIntelBufferLocation attribute, in the second case the _arg_ will the get the RestrictAttr  attribute. 

The types are: cl::sycl::INTEL::property::buffer_location::instance<> and cl::sycl::ONEAPI::property::no_alias::instance<>. 
Can these be added to the enum?

Reply:
I think we could add accessor_property_list to the enum ?

Reply:
I don't think we should use enum for `sycl_special_class` attribute.

> The property_list is only used for the accessor class. Sampler and stream classes don’t have a property_list. It is used to gather information about the kernel arguments. 

Do you mean the compiler is expected to inspect template parameter values for particular instance to gather this information? If so, this sounds like very similar design issue we are trying to address with `sycl_special_class` attribute.

> When appropriate the attributes SYCLIntelBufferLocation and/or RestrictAttr are added to the argument.

I would expect DPC++ runtime library to use attributes explicitly in the source code/headers and rely on the attributes semantics defined by the compiler. E.g. if these are used together with `sycl_special_class`, user can expect that they will be applied to SPIR-V kernel parameters as well.

Reply:
Looking at it a little further, thinking that we could have this attribute:

def SYCLSpecialClass: InheritableAttr {
  let Spellings = [Clang<"sycl_special_class", /*allowInC =*/0>];
  let Subjects = SubjectList<[CXXRecord]>;
  let LangOpts = [SYCLIsDevice, SilentlyIgnoreSYCLIsHost];
  let Documentation = [SYCLSpecialClassDocs];
  **let Args = [IdentifierArgument<"ClassType">,
              TypeArgument<"accessor_property_list", /*optional*/1>];**
}

And in the header have the accessor class be declared this way
typename ONEAPI::accessor_property_list<> propertyListT;
class **__attribute__((sycl_special_class(accessor, propertyListT)))** accessor {
public:
…
};

For sampler and stream the optional accessor_property_list attribute would be null. For accessor the attribute would get the type from the property_list. That would allow having a single type for accessor, sampler and stream classes, but we still have the ability to differentiate the accessor class from the 2 others.

Reply:
@bader said:
> I don't think we should use enum for sycl_special_class attribute.

Can you explain why a bit more?

Reply:
As stated in [the original proposal](https://github.com/intel/llvm/pull/2091/files#r461743627) special types processing is uniform for all standard SYCL classes.

Another reason mentioned in the latest comment from @zahiraam above is to add additional parameters for non-standard extensions support like https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/accessor_properties/, but it's not clear to me if it's necessary or can be avoided. Today we parse instances of SYCL property template to apply corresponding attributes implicitly by the compiler and this approach tries to match the type name defined in DPC++ header i.e. has the same problem we are trying to resolve with this patch. I would like to check if we can use C++ attributes explicitly in DPC++ headers to avoid custom handling of DPC++ implementation internals by the compiler.
For instance, among supported properties there is `noalias`, which I'd like to represent with one of "restrict" attributes.
The trick here is how to define SYCL classes so that template class instantiation has `__init` method declaration with attributes applied corresponding to property_list value.

Reply:
Thank you for the explanation @bader! To make sure I'm actually clear on the proposal though, where we're converging now is:

* We need `sycl_special_class` to identify class types that we expect to see an `__init()` method on that has some special meaning. This attribute does not require any arguments at all, it just tells the "this class is special, look at it harder" and the `__init()` method gives us the extra information we require. We could potentially skip looking at the `__init()` method by using an attribute argument, but because we have to look at that method anyway to validate it, having an attribute argument doesn't buy us much.
* The SYCL headers will apply existing attributes (or newly created) attributes to apply other information to the class, such as `sycl_accessor(property-list)` rather than have the FE infer the attributes from the class.

Have I understood properly? If so, then I think we're in pretty close agreement that it's better to have these attributes written in source under control of the header files rather than implied by the FE based on heuristics. Not only does that simplify the FE, it makes it *far* more clear to anyone reading the header files as to what attributes are in effect for a given declaration.

Reply:
That's right with one minor correction. We allow applying attributes to `__init()` parameters as well and these will be propagated to SPIR kernel parameters to support `no_alias` and `buffer_location` features. For example:
```c++
[[sycl::special_class]] class accessor { // [[sycl_special_class]] or [[sycl::kernel_init]] or some better attribute name...
// ...
void __init(__restrict __global T* p, sycl::range access_size, ...);
```
- We expect compiler to find `__init` member function and clone it's arguments to SPIR kernel arguments with all the attrbutes.
- Apply attributes to `__init` parameters will require some template programming in DPC++ headers, which will parse `properly_list` and apply attributes correspondingly. For the accessor instance like this: `accessor A(B, h, accessor_property_list{no_alias});` DPC++ runtime headers code has to convert `no_alias` accessor property to `__restrict` attribute in `__init` member declaration. According to my understanding this should be possible.

Similar attributes propagation logic can be applied to the attributes applied to SYCL kernel (i.e. lambda or function object). Actually I think we already apply it for attributes, but `property_list` template arguments should be lowered to explicit attributes to free users from using C++ attributes directly.

I think this compiler/runtime interface makes the implementation much more stable to DPC++ changes (including API changes) and keep compiler implementation simple.

BTW, we probably should clarify this in https://github.com/intel/llvm/blob/sycl/sycl/doc/CompilerAndRuntimeDesign.md#lowering-of-lambda-function-objects-and-named-function-objects and https://github.com/intel/llvm/blob/sycl/sycl/doc/KernelParameterPassing.md documents.

Reply:
@bader Thanks.
Let me make sure I understand the proposal.
So far it is clear to me that the accessor class will now have an explicit attribute in the header. 

But I am not clear about the __init method.
Are you proposing that we infer the attributes of the __init method from the accessor class?

Can we not have instead something like this:
 
class **__attribute__((sycl_special_class))** accessor {

public:
  void use(void) const {}
  void use(void *) const {}
  _ImplT<dimensions> impl;

private:
  using PtrType = typename DeviceValueType<dataT, accessTarget>::type *;
  void **__attribute__((property_list(propertyListT)))**  
       __init(PtrType Ptr, range<dimensions> AccessRange,
              range<dimensions> MemRange, id<dimensions> Offset) {}
  friend class stream;
};

Reply:
> Are you proposing that we infer the attributes of the __init method from the accessor class?

Yes, but adding `property_list` attribute doesn't sound appealing to me.
The point of my proposal is move the logic lowering template argument into clang's supported attribute to DPC++ headers.

I'm not sure if we want to have something generic like `property_list` and more over this is very tightly coupled with existing SYCL API. I'd like to decouple runtime library changes from the compiler changes. 

NOTE: I expect that some accessor class properties might not require compiler support and can be handled in DPC++ headers using standard C++ features (e.g. `sycl::property::no_init`, `ONEAPI::property::no_offset`).

Reply:
If we are inferring the __init attribute from the accessor class declaration, then we don’t have explicit attribute of the __init method in the header. Instead we would create a function attribute, and when we find a property list in the class declaration then we set that attribute to the __init method.
So the accessor declaration would be instead something like this:
class **__attribute__((sycl_special_class))** 
      **__attribute__((sycl_accessor(propertyListT)))** accessor {

public:
  void use(void) const {}
  void use(void *) const {}
  _ImplT<dimensions> impl;

private:
  using PtrType = typename DeviceValueType<dataT, accessTarget>::type *;
  void __init(PtrType Ptr, range<dimensions> AccessRange,
              range<dimensions> MemRange, id<dimensions> Offset) {}
  friend class stream;
};

The presence of the sycl_accessor attribute of the accessor class would then imply a “property” attribute of the corresponding __init method.
@Ballman, Aaron That’s feasible right?
We then would have to create 3 attributes: sycl_special_class and sycl_accessor that will be applied explicitly to the  accessor class  and a third one applied to the __init method, that will be inferred from the sycl_accessor.

Reply:
> If we are inferring the __init attribute from the accessor class declaration, then we don’t have explicit attribute of the __init method in the header.

Sorry, if I wasn't clear. I think compiler should not infer any attributes from C++ templates - all attributes must be explicitly applied in the header file as demonstrated in code sample here: https://github.com/intel/llvm/pull/3892#discussion_r663176382.

Only two out of four accessor properties requires additional attribute - `noalias` and `buffer_location`. All other accessor properties are handled by the runtime library.
```C++
class __attribute__((sycl_special_class)) accessor {

public:
  void use(void) const {}
  void use(void *) const {}
  _ImplT<dimensions> impl;

private:
  using PtrType = typename DeviceValueType<dataT, accessTarget, property_list>::type *;
  void __init(PtrType Ptr, range<dimensions> AccessRange,
              range<dimensions> MemRange, id<dimensions> Offset) {}
  friend class stream;
};
```
I expect `PtrType` from the example above apply required attributes explicitly in the source code, so that compiler doesn't need to infer them.